Converter: Enable translation for error messages (#3313)

pwasilewski opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
This contribution refactors error message handling in `Converter` and `DefaultConverter`. The converters now return error message keys and optional arguments. Localization is performed by `MudFormComponent`, which resolves these keys and arguments into localized error messages using the `InternalMudLocalizer` service.
Unfortunately, some method signatures and property types have been modified as part of this change, resulting in breaking changes.
Fixes #3313 

## How Has This Been Tested?

- Visually: by changing the `.resx` file and verifying that the appropriate localized error messages appear.

- Unit Tests: by adding relevant unit tests.

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

pwasilewski committed:
Converter: Enable translation for error messages (#3313)

pwasilewski committed:
fix: using LanguageResource class instead of literals

pwasilewski committed:
fix: Wron mappings

pwasilewski committed:
chore: Adapting and adding extra unit tests

pwasilewski committed:
Merge branch 'MudBlazor:dev' into feat/localize-converter-error-messages

pwasilewski committed:
fix: Fixing SonarQbe warning + build

versile2 commented:
Add these to BindingConverterTests.cs should get you closer to 100% coverage. Mostly just missing Error testing. Which Looks like Converter.cs instead of setting SetErrorMessage and GetErrorMessage directly the error messages should be calling UpdateSetErrorMessage() and UpdateGetErrorMessage() respectively.

```
        [Test]
        public void DefaultConverter_TimeSpan_EdgeCases()
        {
            var converter = new DefaultConverter<TimeSpan>();

            // Test invalid format
            converter.Get("25:00").Should().Be(TimeSpan.Zero); // Invalid hours
            converter.GetError.Should().BeTrue();
            converter.GetErrorMessage.Should().Be("Not a valid time span");

            // Test overflow
            converter.Get("999999:00:00").Should().Be(TimeSpan.Zero);
            converter.GetError.Should().BeTrue();
            converter.GetErrorMessage.Should().Be("Not a valid time span");

            // Test invalid characters
            converter.Get("12:ab:00").Should().Be(TimeSpan.Zero);
            converter.GetError.Should().BeTrue();
            converter.GetErrorMessage.Should().Be("Not a valid time span");

            // Test empty/null
            converter.Get("").Should().Be(TimeSpan.Zero);
            converter.Get(null).Should().Be(TimeSpan.Zero);
        }

        [Test]
        public void DefaultConverter_TimeSpanFormat_CustomFormats()
        {
            var converter = new DefaultConverter<TimeSpan>();

            // Test default format
            var time = new TimeSpan(1, 2, 3);
            converter.Set(time).Should().Be("01:02:03");
            converter.Get("01:02:03").Should().Be(time);

            // Test custom DefaultTimeSpanFormat
            converter.DefaultTimeSpanFormat = @"hh\:mm";
            converter.Set(time).Should().Be("01:02");
            converter.Get("01:02").Should().Be(new TimeSpan(1, 2, 0));

            // Test Format property override
            converter.Format = @"hh\:mm\:ss\.fff";
            var preciseTime = new TimeSpan(0, 1, 2, 3, 456);
            converter.Set(preciseTime).Should().Be("01:02:03.456");
            converter.Get("01:02:03.456").Should().Be(preciseTime);
        }

        [Test]
        public void BaseConverter_Culture_AffectsConversion()
        {
            var cultureInfo = new CultureInfo("fr-FR");
            var converter = new Converter<double, string>
            {
                Culture = cultureInfo,
                SetFunc = (d) => d.ToString(cultureInfo),
                GetFunc = (s) => double.Parse(s, cultureInfo)
            };

            converter.Set(1.5).Should().Be("1,5"); // French uses comma
            converter.Get("1,5").Should().Be(1.5);

            cultureInfo = new CultureInfo("en-US");
            converter = new Converter<double, string>
            {
                Culture = cultureInfo,
                SetFunc = (d) => d.ToString(cultureInfo),
                GetFunc = (s) => double.Parse(s, cultureInfo)
            };

            converter.Set(1.5).Should().Be("1.5"); // US uses period
            converter.Get("1.5").Should().Be(1.5);
        }

        [Test]
        public void BaseConverter_NullFuncs_ReturnsDefaults()
        {
            var converter = new Converter<int, string>();
            converter.Set(42).Should().BeNull(); // SetFunc is null
            converter.Get("42").Should().Be(0);  // GetFunc is null
            converter.SetError.Should().BeFalse();
            converter.GetError.Should().BeFalse();
        }

        [Test]
        public void BaseConverter_OnError_CallbackInvoked()
        {
            // Variable to capture the last error message from OnError callback
            var lastError = string.Empty;

            // Create a converter with intentionally failing SetFunc and GetFunc
            var converter = new Converter<int, string>
            {
                OnError = (msg) => lastError = msg, // Assign OnError callback
                SetFunc = (i) => throw new Exception("Set failed"),
                GetFunc = (s) => throw new Exception("Get failed")
            };

            // Test the Set method
            converter.Set(42);
            lastError.Should().Contain("Set failed");
            converter.SetErrorMessage.Should().Contain("Set failed");
            converter.SetError.Should().BeTrue();

            // Test the Get method
            converter.Get("42");
            lastError.Should().Contain("Get failed");
            converter.GetError.Should().BeTrue();
            converter.GetErrorMessage.Should().Contain("Get failed");
        }
```

ScarletKuro commented:
Thanks for the PR.

I'm not sure I'm a fan of this approach. The design seems off, with attaching the localizer to the Converter from the side.

First, converters are something I plan to rewrite sooner or later. I want to make them more like XAML, where the culture is part of the From/To methods, rather than being just a side property, as we have with the localizer here. Also, I'm not a fan of the universal converter that attempts to handle all conversions. I think the idea was primarily to support inline conversions in Razor syntax, but at this point, I'm not sure it's worth it.

Second, I would approach the translation differently. I wouldn't translate it at the Converter level, so it wouldn't depend on the localizer. Instead, I would spit out the translation keys
```diff
-UpdateGetError(Localizer?[LanguageResource.DefaultConverter_InvalidBoolean] ?? "Not a valid boolean");
+UpdateGetError(LanguageResource.DefaultConverter_InvalidBoolean);
```
And have the components that listen for the `OnError` event of the converters https://github.com/MudBlazor/MudBlazor/blob/6cef7e43db35f9a8beb42e7c767167fbad2f7098/src/MudBlazor/Base/MudFormComponent.cs#L145 handle that. These components would then depend on the localizer, catch the key, and only then perform the translation.
This would feel more natural.

pwasilewski committed:
Merge branch 'MudBlazor:dev' into feat/localize-converter-error-messages

pwasilewski commented:
@ScarletKuro Thank you for the review!

Your solution indeed looks cleaner. However, I have a question regarding handling strings with interpolation. For example:

https://github.com/MudBlazor/MudBlazor/blob/c4f6c7d591af69543197eccd039e4092abab9911/src/MudBlazor/Utilities/BindingConverters/DefaultConverter.cs#L191

https://github.com/MudBlazor/MudBlazor/blob/c4f6c7d591af69543197eccd039e4092abab9911/src/MudBlazor/Utilities/BindingConverters/DefaultConverter.cs#L196

I initially considered changing
https://github.com/MudBlazor/MudBlazor/blob/c4f6c7d591af69543197eccd039e4092abab9911/src/MudBlazor/Utilities/BindingConverters/Converter.cs#L74
 to `UpdateGetError(string msg, params object[] arguments)` to support interpolation, but this would also require modifying 

https://github.com/MudBlazor/MudBlazor/blob/c4f6c7d591af69543197eccd039e4092abab9911/src/MudBlazor/Utilities/BindingConverters/Converter.cs#L29

to something like a `Tuple`. This approach doesn’t feel entirely appropriate either. Do you have any recommendations for handling this scenario?

Alternatively, I could change the existing behavior in
https://github.com/MudBlazor/MudBlazor/blob/6cef7e43db35f9a8beb42e7c767167fbad2f7098/src/MudBlazor/Base/MudFormComponent.cs#L174

and implement a custom solution, but that also seems strange.

ScarletKuro commented:
I guess we could do a breaking change and make it
```
public Action<string, object[]>? OnError { get; set; }
public (string?, object[]) GetErrorMessage { get; set; }
protected void UpdateGetError(string msg, object[] arguments)
```
?

pwasilewski committed:
Merge branch 'MudBlazor:dev' into feat/localize-converter-error-messages

pwasilewski committed:
fix: Apply ScarletKuro's feedback and reorganize BindingConverterTests unit tests.

pwasilewski committed:
Merge branch 'feat/localize-converter-error-messages' of https://github.com/pwasilewski/MudBlazor into feat/localize-converter-error-messages

pwasilewski committed:
fix: code quality

pwasilewski commented:
Feedback applied. I hope it's better now.

I also made two additional changes since the last update:
- I changed the suffix for the keys from _DefaultConverter_ to _Converter_ to make it more general.
- I took inspiration from versile2's suggestion and reorganized the unit tests in `BindingConverterTests`. Instead of having one method for the happy cases and one for negative cases, I split the tests by type. (I hope this change will help a bit for the future reworks of the converters.)

pwasilewski committed:
Merge branch 'dev' into feat/localize-converter-error-messages

ScarletKuro reviewed (approved):
LGTM

ScarletKuro committed:
Merge branch 'dev' into feat/localize-converter-error-messages

henon commented:
@ScarletKuro As you can see from the issue number, this issue has existed for a long time. I know this is needed, so I'd say let's go for it. 

I know that @HClausing is using converters. @HClausing what do you think about merging this in v8? 

@pwasilewski can you write a sentence for our migration guide that advises users how they should adapt their code to this breaking change?

pwasilewski commented:
@henon 
Sure! Does something like this work?

In **Converter.cs**, several property signatures have been updated to enhance localization support. These properties no longer return English messages directly. Instead, they now return keys intended for localization, accompanied by additional parameters:

- **OnError**: Updated from `Action<string>?` to `Action<string, object[]>?`. This allows passing a key and optional parameters, enabling more flexible and accurate localization.
- **SetErrorMessage** and **GetErrorMessage**: Both changed from `string?` to `(string, object[])?`. These properties now provide a key for localization along with relevant parameters, shifting message formatting to the localization system.

pwasilewski committed:
Merge branch 'dev' into feat/localize-converter-error-messages

ScarletKuro commented:
Added to v8.0.0 Migration Guide https://github.com/MudBlazor/MudBlazor/issues/9953

HClausing commented:
> @ScarletKuro As you can see from the issue number, this issue has existed for a long time. I know this is needed, so I'd say let's go for it.
> 
> I know that @HClausing is using converters. @HClausing what do you think about merging this in v8?
> 
> @pwasilewski can you write a sentence for our migration guide that advises users how they should adapt their code to this breaking change?

I really like this approach.

Yes, I've written some custom Converters to dynamically apply masks for brazilian social ID's (companies and natural peopla have different masks, but should be placed on the same field on model/binding).
 
Would be nice to have it on v8.

henon commented:
Thanks @pwasilewski !
